storm-client/src/jvm/org/apache/storm/executor/ExecutorTransfer.java

```diff
@@ -17,72 +17,92 @@
         this.workerData = workerData;
-        this.batchTransferQueue = batchTransferQueue;
-        this.topoConf = topoConf;
         this.serializer = new KryoTupleSerializer(topoConf, workerData.getWorkerTopologyContext());
-        this.cachedEmit = new MutableObject(new ArrayList<>());
         this.isDebug = ObjectReader.getBoolean(topoConf.get(Config.TOPOLOGY_DEBUG), false);
+        this.producerBatchSz = ObjectReader.getInt(topoConf.get(Config.TOPOLOGY_PRODUCER_BATCH_SIZE));
+        this.localReceiveQueues = Utils.convertToArray(workerData.getShortExecutorReceiveQueueMap());
+        this.outboundQueues = new ArrayList<JCQueue>(Collections.nCopies(localReceiveQueues.size(), null) );
     }
 
-    public void transfer(int task, Tuple tuple) {
-        AddressedTuple val = new AddressedTuple(task, tuple);
+
+    public void transfer(int task, Tuple tuple) throws InterruptedException {
+        AddressedTuple addressedTuple = new AddressedTuple(task, tuple);
         if (isDebug) {
-            LOG.info("TRANSFERRING tuple {}", val);
+            LOG.info("TRANSFERRING tuple {}", addressedTuple);
         }
-        batchTransferQueue.publish(val);
-    }
 
-    @VisibleForTesting
-    public DisruptorQueue getBatchTransferQueue() {
-        return this.batchTransferQueue;
+        boolean isLocal = transferLocal(addressedTuple);
+        if (!isLocal) {
+            transferRemote(addressedTuple);
+            ++remotesBatchSz;
+            if(remotesBatchSz >=producerBatchSz) {
+                flushRemotes();
+                remotesBatchSz =0;
+            }
+        }
     }
 
-    @Override
-    public Object call() throws Exception {
-        batchTransferQueue.consumeBatchWhenAvailable(this);
-        return 0L;
+    private void transferRemote(AddressedTuple tuple) throws InterruptedException {
+        workerData.transferRemote(tuple);
     }
 
-    public String getName() {
-        return batchTransferQueue.getName();
+    // flushes local and remote messages
+    public void flush() throws InterruptedException {
+        flushLocal();
+        flushRemotes();
     }
 
-    @Override
-    public void onEvent(Object event, long sequence, boolean endOfBatch) throws Exception {
-        ArrayList cachedEvents = (ArrayList) cachedEmit.getObject();
-        cachedEvents.add(event);
-        if (endOfBatch) {
-            workerData.transfer(serializer, cachedEvents);
-            cachedEmit.setObject(new ArrayList<>());
+    private void flushLocal() throws InterruptedException {
+        for (int i = 0; i < outboundQueues.size(); i++) {
+            JCQueue q = outboundQueues.get(i);
+            if(q!=null)
+                q.flush();
         }
     }
+
+    private void flushRemotes() throws InterruptedException {
+        workerData.flushRemotes();
+    }
+
+    public boolean transferLocal(AddressedTuple tuple) throws InterruptedException {
+        workerData.checkSerialize(serializer, tuple);
```

Comment:
I guess this should be done only when it is determined to be sending to local task. So, this should be pushed to just before `queue.publish(tuple)` below.